src/challenge-2/challenge-2.c

```diff
@@ -1,29 +1,89 @@
-#include <json-c/json.h>
-#include <stdio.h>
-
 #include "../../lib/collections.h"
 #include "../../lib/util.h"
 #include "json-c/json_object.h"
+#include <json-c/json.h>
+#include <stdio.h>
+
+char* LEADER_NODE = "n0";
 
 json_object* init_reply(json_object* init_msg);
 
 void client_request_handler(json_object* client_request);
 void conch_response_handler(json_object* conch_response);
+void conch_request_handler(json_object* conch_request, Queue* conch_queue);
+void conch_release_handler(json_object* conch_release, Conch* conch);
+void conch_response_dispatcher(Conch* conch, Queue* request_queue);
 
-char* LEADER_NODE = "n0";
-
-void client_request_handler(json_object* client_request)
+int main(void)
 {
-
-    /* Conch request schema
+    json_object* init_msg = msg_recv();
+    if (init_msg == NULL)
     {
-      "src": "n_",
-      "dest": "n0",
-      "body": { "type": "conch_request", "client_id": "c_" }
+        fprintf(stderr, "expected init message, got EOF");
+        exit(EXIT_FAILURE);
     }
-    */
+    msg_send(init_reply(init_msg));
+    json_object_put(init_msg);
+
+    // TODO: Event loops
+}
+
+json_object* init_reply(json_object* init_msg)
+{
+    return generic_reply(init_msg);
+}
+
+// Used exclusively by Leader
+void conch_request_handler(json_object* conch_request, Queue* request_queue)
+{
+    json_object* body = json_object_object_get(conch_request, "body");
+    queue_enqueue(request_queue,
+                  json_object_object_get(body, "original_client_request"));
+}
+
+// Used exclusively by Leader
+void conch_response_dispatcher(Conch* conch, Queue* request_queue)
+{
+    json_object* conch_response = json_object_new_object();
+    json_object* original_client_request = queue_dequeue(request_queue);
+
+    // Construct conch response message
+    json_object* src = json_object_new_string(LEADER_NODE);
+    json_object_object_add(conch_response, "src", src);
+
+    json_object* dest = json_object_object_get(original_client_request, "dest");
+    json_object_get(dest);
+    json_object_object_add(conch_response, "dest", dest);
+
+    json_object* body = json_object_new_object();
+    json_object_object_add(body, "type",
+                           json_object_new_string("conch_response"));
+    int64_t conch_value = conch_checkout(conch);
+    json_object_object_add(body, "conch_value",
+                           json_object_new_int64(conch_value));
+    json_object_get(original_client_request);
```

Comment:
This is subtle, but I think you actually don't want the `_get` here.

Say `original_client_response` has refcount 1 when you dequeue it. You're gonna stuff it in the `conch_response` as a field, but you want it *stay at* refcount 1, so that it gets freed whenever the `conch_response` does.

Loosely speaking, the "ownership" of the `original_client_response` object is being transferred from the `Queue` to the `conch_response` -- so there was 1 owner before, and 1 owner after -- no refcount updates required.